src/Sentry/Internal/Http/CachingTransport.cs

```diff
@@ -254,17 +257,29 @@ private void EnsureFreeSpaceInCache()
 
     private async Task ProcessCacheAsync(CancellationToken cancellation)
     {
-        // Signal that we can start waiting for _options.InitCacheFlushTimeout
-        _preInitCacheResetEvent?.Set();
+        try
+        {
+            // Make sure we're the only thread processing items
+            await _processingSignal.WaitAsync(cancellation).ConfigureAwait(false);
 
-        // Process the cache
-        while (await TryPrepareNextCacheFileAsync(cancellation).ConfigureAwait(false) is { } file)
+            // Signal that we can start waiting for _options.InitCacheFlushTimeout
+            _preInitCacheResetEvent?.Set();
+
+            // Process the cache
+            _options.LogDebug("Flushing cached envelopes.");
+            while (await TryPrepareNextCacheFileAsync(cancellation).ConfigureAwait(false) is { } file)
```

Comment:
That means whoever - flush or cache worker -goes first, processes **all** the cached envelopes, because in the end, we don't really care who sends them?

Reply:
Correct.  Whichever is in progress should continue to send.  In most cases then, flushing now is really just blocking the calling thread until the background thread has completed.